modules/navier_stokes/src/fvkernels/CNSFVHLLCBase.C

```diff
@@ -0,0 +1,119 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "CNSFVHLLCBase.h"
+#include "NS.h"
+#include "HLLCUserObject.h"
+#include "SinglePhaseFluidProperties.h"
+
+namespace nms = NS;
+using MetaPhysicL::raw_value;
+
+std::vector<std::unordered_map<const FaceInfo *, std::array<ADReal, 3>>>
+    CNSFVHLLCBase::_fi_to_wave_speeds;
+
+InputParameters
+CNSFVHLLCBase::validParams()
+{
+  InputParameters params = FVFluxKernel::validParams();
+  params.addRequiredParam<UserObjectName>(nms::fluid, "Fluid userobject");
+  return params;
+}
+
+CNSFVHLLCBase::CNSFVHLLCBase(const InputParameters & params)
+  : FVFluxKernel(params),
+    _fluid(dynamic_cast<FEProblemBase *>(&_subproblem)
+               ->getUserObject<SinglePhaseFluidProperties>(nms::fluid)),
+    _specific_internal_energy_elem(getADMaterialProperty<Real>(nms::specific_internal_energy)),
+    _specific_internal_energy_neighbor(
+        getNeighborADMaterialProperty<Real>(nms::specific_internal_energy)),
+    _rho_et_elem(getADMaterialProperty<Real>(nms::total_energy_density)),
+    _rho_et_neighbor(getNeighborADMaterialProperty<Real>(nms::total_energy_density)),
+    _vel_elem(getADMaterialProperty<RealVectorValue>(nms::velocity)),
+    _vel_neighbor(getNeighborADMaterialProperty<RealVectorValue>(nms::velocity)),
+    _speed_elem(getADMaterialProperty<Real>(nms::speed)),
+    _speed_neighbor(getNeighborADMaterialProperty<Real>(nms::speed)),
+    _rho_elem(getADMaterialProperty<Real>(nms::density)),
+    _rho_neighbor(getNeighborADMaterialProperty<Real>(nms::density)),
+    _pressure_elem(getADMaterialProperty<Real>(nms::pressure)),
+    _pressure_neighbor(getNeighborADMaterialProperty<Real>(nms::pressure))
+{
+  if (_tid == 0)
```

Comment:
`_fi_to_wave_speeds` is a static data member so every thread will share the same `_fi_to_wave_speeds` data structure. this is similar to RC coeffs (currently). Before I took up the compressible work the `waveSpeed` method was already static, but there was no caching so mass, all momentum components, and energy would trigger redundant wave speed computations on a per FaceInfo basis. I added this static data member to prevent that